src/main/java/com/instaclustr/cassandra/driver/auth/KerberosAuthProvider.java

```diff
@@ -270,45 +294,51 @@ private static Subject loginAsSubject()
         }
 
         @Override
-        public byte[] initialResponse()
+        public CompletionStage<ByteBuffer> initialResponse()
         {
 
             if (saslClient.hasInitialResponse())
             {
                 try
                 {
-                    return Subject.doAs(subject, (PrivilegedExceptionAction<byte[]>) () ->
+                    byte[] response = Subject.doAs(subject, (PrivilegedExceptionAction<byte[]>) () ->
                             saslClient.evaluateChallenge(new byte[0]));
+                    return CompletableFuture.completedFuture(ByteBuffer.wrap(response));
                 } catch (PrivilegedActionException e)
                 {
                     throw new RuntimeException(e.getException());
                 }
             }
 
-            return new byte[0];
+            return CompletableFuture.completedFuture(ByteBuffer.wrap(new byte[0]));
         }
 
         @Override
-        public byte[] evaluateChallenge(final byte[] challenge) {
-
+        public CompletionStage<ByteBuffer> evaluateChallenge(@Nullable ByteBuffer challenge) {
             try
             {
-                return Subject.doAs(subject, (PrivilegedExceptionAction<byte[]>) () ->
-                        saslClient.evaluateChallenge(challenge));
+                byte[] bytes = new byte[challenge.capacity()];
```

Comment:
There's a potential NPE here. With challenge marked as nullable I'd argue this should not just accept the NPE. We could pass on an empty challenge to the client but empty is not the same as null semantically so I'm thinking just return null to assert no further action required and log about this.
```suggestion
            if (challenge == null) {
                logger.warn("Cannot evaluate a null challenge, skipping");
                return CompletableFuture.completedFuture(null);
            }
```